lib/src/front_end/delimited_list_builder.dart

```diff
@@ -83,27 +123,68 @@ class DelimitedListBuilder {
     }
   }
 
-  /// Adds any comments preceding [token] the list.
+  /// Inserts an inner left delimiter between two elements.
+  ///
+  /// This is used for parameter lists when there are both mandatory and
+  /// optional or named parameters to insert the `[` or `{`, respectively.
+  ///
+  /// This should not be used if [delimiter] appears before all elements. In
+  /// that case, pass it to [leftBracket].
+  void leftDelimiter(Token delimiter) {
+    assert(_elements.isNotEmpty);
+
+    // Preserve any comments before the delimiter. Treat them as occurring
+    // before the previous element's comma. This means that:
+    //
+    // ```
+    // function(p1, /* comment */ [p1]);
+    // ```
+    //
+    // Will be formatted as:
+    //
+    // ```
+    // function(p1 /* comment */, [p1]);
+    // ```
+    //
+    // (In practice, it's such an unusual place for a comment that it doesn't
+    // matter that much where it goes and this seems to be simple and
+    // reasonable looking.)
+    _commentsBeforeComma = _commentsBeforeComma
+        .concatenate(_visitor.takeCommentsBefore(delimiter));
+
+    // Attach the delimiter to the previous element.
+    _elements.last = _elements.last.withDelimiter(delimiter.lexeme);
+  }
+
+  /// Adds [comments] to the list.
   ///
   /// If [hasElementAfter] is `true` then another element will be written after
   /// these comments. Otherwise, we are at the comments after the last element
   /// before the closing delimiter.
-  void _addComments(Token token, {required bool hasElementAfter}) {
-    var commentsBeforeElement = _visitor.takeCommentsBefore(token);
-
+  void _addComments(CommentSequence comments, {required bool hasElementAfter}) {
     // Early out if there's nothing to do.
-    if (_commentsBeforeComma.isEmpty && commentsBeforeElement.isEmpty) return;
+    if (_commentsBeforeComma.isEmpty && comments.isEmpty) return;
 
     // Figure out which comments are anchored to the preceding element, which
     // are freestanding, and which are attached to the next element.
     var (
+      inline: inlineComments,
       hanging: hangingComments,
       separate: separateComments,
       leading: leadingComments
-    ) = _splitCommaComments(commentsBeforeElement,
-        hasElementAfter: hasElementAfter);
+    ) = _splitCommaComments(comments, hasElementAfter: hasElementAfter);
 
-    // Add any hanging comments to the previous argument.
+    // Add any hanging inline block comments to the previous element before the
+    // subsequent ",".
+    if (inlineComments.isNotEmpty) {
+      for (var comment in inlineComments) {
```

Comment:
The for-in loop should already handle `isEmpty` input. Can we avoid the extra condition and indentation?

Reply:
Good catch, yes. (This was copied from the subsequent section for `hangingComments` which has some extra code outside of the loop.)